File: Looped-BST/bst_node.m
#import <Foundation/Foundation.h>
#import <QuartzCore/QuartzCore.h>

@interface Node:NSObject{
  Node *right;
  Node *left;
  Node *parent;
  int value;
}

-(Node *) insertValue: (Node *) root Value:(int) valueToInsert;
-(Boolean) searchForValue: (Node *) root Value:(int) valueToSearch;

@property (assign) Node *left;
@property (assign) Node *right;
@property (assign) Node *parent;

@end

@implementation Node

@synthesize left;
@synthesize right;
@synthesize parent;

-(Node *) insertValue:(Node *) root Value:(int) valueToInsert{

  Node *newNode = [[Node alloc] init];
  newNode->value = valueToInsert;
  newNode.left = nil;
  newNode.right = nil;
  newNode.parent = nil;

  Node *temp = root;

  if(root == nil){
    //printf("The root is %d.\n",valueToInsert);
    root = newNode;
  }
  else{
    while (temp != nil){
      if (temp->value > valueToInsert){
        //printf("%d is in the left of %d\n",valueToInsert,temp->value);
        //navigate the left side
        if(temp.left == nil){
          //printf("%d is in the left of %d\n",valueToInsert,temp->value);
          newNode.parent = temp;
          temp.left = newNode;
          break;
        }
        else{
          temp = temp.left;
          continue;
        }
      }
      else if(temp->value < valueToInsert){
        //printf("%d is in the right of %d\n",valueToInsert,temp->value);
        //navigate the right side
        if(temp.right == nil){
          //printf("%d is in the right of %d\n",valueToInsert,temp->value);
          newNode.parent = temp;
          temp.right = newNode;
          break;
        }
        else{
          temp = temp.right;
          continue;
        }
      }
    }
  }
  return root;
}

-(Boolean) searchForValue: (Node *) root Value:(int) valueToSearch{

  Node *temp = root;

  if(root == nil){
    return false;
  }

  else {
    while (temp != nil){
      if(temp->value == valueToSearch){
        return true;
      }
      else if (temp->value > valueToSearch){
        if(temp.left == nil){
          return false;
        }
        else {
          temp = temp.left;
          continue;
        }
      }
      else if(temp->value < valueToSearch){
        if(temp.right == nil){
          return false;
        }
        else {
          temp = temp.right;
          continue;
        }
      }
    }
  }
  return false;
}
@end

File: Looped-BST/bst_node.swift
import Foundation
import QuartzCore

class Node {
  var value:Int?
  var right:Node?
  var left:Node?
  var parent:Node?

  init(){
    self.right = nil
    self.left = nil
    self.parent = nil
    //self.value = nil
  }

  func insertValue(root:Node?, valueToInsert:Int) -> Node {
    var newNode = Node()
    newNode.value = valueToInsert

    var temp = root

    if (root == nil) {
      print("\(valueToInsert) is the root")
      return newNode
    }
    else{
      while (temp != nil) {
        if var tempVal = temp?.value{
          if(tempVal > valueToInsert){
            if(temp?.right == nil){
              print("\(valueToInsert) is on the right of \(tempVal)")
              newNode.parent = temp
              temp?.right = newNode
              break
            }
            else{
              temp = temp?.right
            }
          }
          else if(tempVal < valueToInsert){
            if(temp?.left == nil){
              print("\(valueToInsert) is on the left of \(tempVal)")
              newNode.parent = temp
              temp?.left = newNode
              break
            }
            else{
              temp = temp?.left
            }
          }
        }
        }
      while(temp?.parent != nil){
        temp = temp?.parent
      }
    }
    return temp!
  }
}
